consensus/clique/src/main/java/tech/pegasys/pantheon/consensus/clique/CliqueExtraData.java

```diff
@@ -47,9 +56,28 @@ public CliqueExtraData(
     this.vanityData = vanityData;
     this.proposerSeal = Optional.ofNullable(proposerSeal);
     this.validators = validators;
+    proposerAddress =
+        Suppliers.memoize(() -> CliqueBlockHashing.recoverProposerAddress(header, this));
+  }
+
+  public static BytesValue createUnsignedData(
+      final BytesValue vanityData, final List<Address> validators) {
+    return new CliqueExtraData(vanityData, null, validators, null).encode();
+  }
+
+  public static CliqueExtraData decode(final BlockHeader header) {
+    final Object inputExtraData = header.getParsedExtraData();
+    if (inputExtraData instanceof CliqueExtraData) {
+      return (CliqueExtraData) inputExtraData;
+    }
+    LOG.warn(
```

Comment:
I would expect to never see this message because the extra data should only ever be parsed to the right type.  But the nice thing about this approach is that if somehow we get that wrong, things we still work, just with a performance penalty (and this warning is intended to make that situation visible).